feat(challenge): allow user to save partial solution and continue later (#547)

Silvitui opened the pull request:
### 🧩 Context

This PR implements the functionality for users to save a **partial solution** while working on a challenge, and resume it later without losing progress. Defined in task #547 User story : #541.
### ⚠️ Note

The branch name is incorrect — it should be `feat/547-save-partial-solution` instead of `feat/541-save-partial-solution`.  

---

### ✅ What’s included

- 💾 **Partial solution saving**: Users can now write part of their solution and click "SAVE" to store their progress. The solution is saved in the database with status `'IN_PROGRESS'`.

- 🔁 **Challenge state detection**:
  - If a challenge is **not started**, the button **"Iniciar repte"** is shown.
  - If it was already started but not finalized (`IN_PROGRESS`), the user sees **"Continuar repte"**, along with **"Guardar"** and **"Finalitzar"** buttons.
  - If the challenge was **finalized** (`ENDED`), the user's solution is shown but cannot be edited.

- 🧼 **Reset editor content properly**:
  - If a user enters a challenge that has **no saved solution**, the editor now **starts empty**.
  - This prevents showing stale content from a previous challenge.

- 👤 **Role-based UI**:
  - Admin users do not see challenge action buttons (start/save/finish), only students (USER role) do.

- ⚙️ **Refactored logic** to ensure that solution state (`IN_PROGRESS`, `ENDED`) is correctly set based on backend data.

---

### 🧾 Technical note / Backend alignment required

To fully support this frontend feature, the backend should add support for a new solution status: `"NOT_STARTED"`.

Currently, the frontend handles this as a **UI-only state** to show the “Start challenge” button when no solution exists yet. However, to ensure full synchronization between frontend and backend, this state should ideally be recognized on the backend as well.

**Recommendation:**  
Extend the backend `UserSolutionStatus` enum (or equivalent logic) to include:


NOT_STARTED = 'NOT_STARTED'

### 🧪 How to test

1. Open any challenge as a **USER**, click **"Iniciar repte"**.
2. Type code like `console.log("test")`, click **"Guardar"**.
3. Navigate to a different challenge:
   - If it has no previous solution, the editor should be **empty**.
   - If it has a saved solution (`IN_PROGRESS`), it should load that content.
4. Navigate back to the original challenge → the previously saved code should appear.
5. Finalize a challenge and ensure the buttons disappear and the solution becomes read-only.
6. Log in as **ADMIN** and confirm that challenge action buttons are not displayed.

---
---

---

### 🖼️ Screenshots

#### ✅ Solution already submitted
Here we have the view after the solution has been finalized. The challenge is marked as completed, and no editing is possible.

![Solution submitted](https://github.com/user-attachments/assets/849a8576-4ae5-49f6-8c4e-62580f174e6a)

---

#### 🔄 Continue challenge
This screen appears when the user has previously saved a partial solution. The "Continue Challenge" button allows the user to resume where they left off.

![Continue challenge](https://github.com/user-attachments/assets/9b155e06-0b69-4735-aaee-de51e9ebca84)

---

#### 💾 Save or finalize
After clicking "Continue", the user sees the editor along with the "Save" and "Finish" buttons, giving them the choice to continue working or submit their final solution.

![Save or finalize](https://github.com/user-attachments/assets/8fe35a55-54c2-42c6-a078-6b016e86776c)


#### MODIFIED FILES SUMMARY

#### `user-solution.interface.ts`
- ➕ Added `status` field to track the current state of the user’s solution (`IN_PROGRESS` or `ENDED`).
- 🎯 Enables partial saving and distinguishes between a draft and a finalized challenge.

#### `challenge-header.component.ts`
- 🧠 Detects challenge state dynamically: `NOT_STARTED`, `IN_PROGRESS`, `ENDED`.
- 🔄 Triggers reset of the editor when switching to a new challenge without saved text.
- 👤 Controls UI based on user role (ADMIN or USER) and challenge state.

#### `challenge-header.component.html`
- 👀 Shows/hides action buttons (`Start`, `Continue`, `Save`, `Finish`) based on the current solution state.
- 🔒 Prevents showing mentor’s solution until the user has finalized the challenge.

#### `editor-challenge.component.ts`
- 🧩 Receives solution text via `@Input()` and uses `ngOnChanges` to update content reactively.
- ✍️ Emits updated text to parent via `@Output()` every time the editor changes.
- 🔁 Resets textarea correctly when switching to a challenge with no saved solution.

#### `challenge-info.component.ts`
- 🔄 Manages the current state of the solution and toggles tab visibility.
- 🧠 Keeps logic centralized ).
- 👂 Listens to header and editor component events.

#### `challenge-info.component.html`
- 📌 Controls visibility of tabs and messages depending on solution state:
  - ❌ No tab shown if not started.
  - ✍️ Shows only user’s solution if `IN_PROGRESS`.
  - ✅ Shows both user and mentor solutions if `ENDED`.
  

  - #### `editor-challenge.component.ts`
- ❌ Removed localStorage auto-save on editor content changes.  
- 🔄 Now emits content changes via events and updates service only.


#### `modals/send-solution-modal/send-solution-modal.component.ts`
- 📤 Handles `END` status update on challenge finalization.
- 🧭 Ensures proper solution submission flow and modal behavior post-submit.

#### `modals/send-solution-modal/send-solution-modal.component.html`
- ✅ Confirms submission visually.
Removed localStorage auto-save from editor

###  The editor no longer saves content automatically to localStorage on changes. Content is emitted and managed via events and services only.


### ✅ Why was this test removed?
We deleted the logic inside `onStartChallenge()` that stored data in `localStorage`, like:

```ts
localStorage.setItem('challengeStarted', JSON.stringify({ id: this.idChallenge, started: true }));
localStorage.setItem('currentChallengeId', this.idChallenge);
this.startChallenge.emit(true);
await this.router.navigate(['/ita-challenge/challenges', this.idChallenge, 'start']);

Now, we don’t need to use it...Keeping the localStorage logic caused problems, because we had two sources of truth:
localStorage
Backend 
This sometimes made the "Start Challenge" button appear twice or at the wrong time.

Silvitui committed:
fix: show/hide challenge action buttons based on user role and state

Silvitui committed:
fix: show/hide challenge action buttons based on user role and state

Silvitui committed:
fix(challenge): reset solutionText when switching challenges without a saved solution to avoid showing previous answer

Silvitui committed:
fix(challenge): reset solutionText when switching challenges without a saved solution to avoid showing previous answer

Silvitui committed:
✅ Fixed failing tests

Silvitui committed:
add CHANGELOG

Silvitui committed:
add CHANGELOG

Silvitui committed:
test(challenge-info): add unit tests to increase coverage and pass SonarQube Quality Gate

oscarrep commented:
I have tested these changes and confirmed they are working as intended: as USER i can save progress on a challenge and come back to it later to finish it, and as ADMIN i can't see the challenge action buttons.

Great work overall! Let's hope you can pass the Sonar checks 🙏

Silvitui committed:
adding test to pass the coverage sonarqube

Silvitui committed:
updating changelog

Silvitui committed:
Save local changes in package-lock.json before merge

Silvitui committed:
Remove accidental translation key and restore i18n files to original state

ivilarop commented on et --hard ae3c640f:
what is this file?

Silvitui replied:
I’ve already updated the changelog to reflect it. Let me know if anything else is missing

ivilarop commented:
> * NOT_STARTED

We don't have a NOT_STARTED status in the backend. What should we do?

<img width="841" height="49" alt="image" src="https://github.com/user-attachments/assets/4e0dc065-f5c6-48bc-8490-73dca085c479" />

<img width="697" height="138" alt="image" src="https://github.com/user-attachments/assets/4a75dd88-8543-468a-bacc-584c62c8623d" />

ivilarop commented on src/app/models/user-solution.interface.ts:
```diff
@@ -3,6 +3,7 @@ export interface UserSolution {
   uuid_challenge: string
   uuid_language: string
   solution_text: string
+  status: 'IN_PROGRESS' | 'ENDED'
```
I see that in user-solution.interface.ts you've declared the status type as a union of literals 'IN_PROGRESS' | 'ENDED'.
Why not just reuse the SolutionStatus enum you just created?
This way we maintain a single source of truth for the statuses and avoid duplicate strings.

Silvitui replied:
Yes..it's better to use the SolutionStatus enum. I've changed it now in the interface. Thanks for the suggestion

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.html:
```diff
@@ -49,8 +49,11 @@ <h3 class="ms-2">{{title | dynamicTranslate}}</h3>
   <div class="col-md-4 button d-none d-md-flex">
     <ng-container *ngIf="!solutionSent; else solutionSentMessage">
+       <button
+      *ngIf="solutionState === 'IN_PROGRESS' && !challengeStarted && userRole !== USER_ROLE.ADMIN"
```
In *ngIf, you're using 'IN_PROGRESS' as a literal string. Since you've already created the SolutionStatus enum, could you reuse it here as well? This way, we avoid typos and maintain consistency throughout the app.

Also, I see that in *ngIf you're mixing solutionState === 'IN_PROGRESS' with !challengeStarted.

Does it make sense for them to coexist in validation?

Silvitui replied:
Thanks! You're right. I’ll use the SolutionStatus enum instead of the string.
About the second part, I’ll check the logic to see if we really need both conditions.

ivilarop replied:
About "As for the second part, I'll review the logic to see if we really need both conditions."

What conclusion did you reach?

Silvitui replied:
I added the condition !challengeStarted because without it, the "Continue challenge" button is always visible.. even when the challenge has already started.
This condition hides the button when the challenge is running, and only shows it when the user hasn’t started it yet but has a saved draft.
In the future, we can refactor the logic to improve this

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.html:
```diff
@@ -49,8 +49,11 @@ <h3 class="ms-2">{{title | dynamicTranslate}}</h3>
+class="btn btn-sm btn-primary"(click)="onContinueChallenge()">{{ 'modules.challenge.header.buttonContinueChallenge' | translate }}</button>
   <button
-  *ngIf="!challengeStarted && userRole !== USER_ROLE.ADMIN" class="btn btn-sm btn-primary" (click)="onStartChallenge()" >
+  *ngIf="solutionState === 'NOT_STARTED' && userRole !== USER_ROLE.ADMIN" class="btn btn-sm btn-primary" (click)="onStartChallenge()" >
```
The literal 'NOT_STARTED' appears in the template, but this value doesn't exist in the SolutionStatus enum or in the API.

What should we do?

Silvitui replied:
You're right ..'NOT_STARTED' is not part of the backend or the SolutionStatus enum.
We added it locally as part of the SolutionState type to handle the case when there's no solution yet.
It helps us manage the UI before the user starts a challenge, without needing extra values from the backend.

Silvitui commented:
We don't have a NOT_STARTED status in the backend. But in the frontend, we added "NOT_STARTED" as a local UI state for a specific reason:
It helps us detect when the user hasn't started the challenge yet, meaning there is no solution stored in the database for that challenge and user. Based on that, we can display the "Start challenge" button instead of "Continue challenge" or "Challenge completed".
So it's not part of the backend model. It's just a derived state used on the client side to improve the user experience. 
@ivilarop 
I’ve updated the description to clarify that the logic was refactored to rely only on the backend solution status (IN_PROGRESS, ENDED).

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -40,69 +41,84 @@ export class ChallengeHeaderComponent implements OnInit {
   @Output() favoritesUpdated = new EventEmitter<number>()
+  @Input() solutionText: string = '';
+  @Input() status: string = '';
+  @Input() solutionState: 'NOT_STARTED' | 'IN_PROGRESS' | 'ENDED' = 'NOT_STARTED';
```
Magic Strings, Reuse the existing SolutionStatus enum

Silvitui replied:
i’ve updated the solutionState to use the SolutionStatus enum and removed the magic strings.

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -40,69 +41,84 @@ export class ChallengeHeaderComponent implements OnInit {
-    // Verifica si el reto ya ha comenzado
-    if (this.challengeStarted) {
-      this.activeId = ChallengeTab.SOLUTIONS
+        if (solution.status === 'IN_PROGRESS') {
```
This block can be refactored to gain readability, avoid duplication, and correctly use the SolutionStatus enum.

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -40,69 +41,84 @@ export class ChallengeHeaderComponent implements OnInit {
-    this.authService.getUserId().subscribe(userId => {
-      this.userId = userId;
+    if (!userId) {
+      console.error("Could not get User ID");
```
What happens after console.error? Does the flow continue to run? 

Would the if else be necessary after the refactor?

Silvitui replied:
Yes! I think  (!userId) return is better than using if/else here.
It's shorter and easier to read. we just stop the code early if there is no user ID.

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -40,69 +41,84 @@ export class ChallengeHeaderComponent implements OnInit {
 
   async onStartChallenge (): Promise<void> {
     this.challengeStarted = true
+    this.solutionState = 'IN_PROGRESS';
```
enum?

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -124,6 +140,9 @@ export class ChallengeHeaderComponent implements OnInit {
     })
     modalRef.componentInstance.idChallenge = this.idChallenge;
     modalRef.componentInstance.userId = this.userId;
+    modalRef.componentInstance.status = 'ENDED';
```
Magic String. enum?

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -151,10 +170,28 @@ export class ChallengeHeaderComponent implements OnInit {
+    return;
   }
 
+  this.status = 'IN_PROGRESS';
```
Magic String

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.spec.ts:
```diff
@@ -40,12 +40,16 @@ describe('ChallengeHeaderComponent', () => {
             uuid_challenge: "testChallengeId",
+            uuid_language: "testLang",
             solution_text: "some solution",
+            status: "ENDED",
```
magic String

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -151,10 +170,28 @@ export class ChallengeHeaderComponent implements OnInit {
-    console.log('Guardando reto...')
+saveChallenge(): void {
+  if (!this.idChallenge || !this.languageId || !this.solutionText || !this.userId) {
+    console.error(' Faltan datos para guardar la solución');
```
text in english

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -151,10 +170,28 @@ export class ChallengeHeaderComponent implements OnInit {
+      this.solutionState = 'IN_PROGRESS';
+    },
+    error: (err) => {
+      console.error(' Error al guardar solución', err);
```
text in english

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -151,10 +170,28 @@ export class ChallengeHeaderComponent implements OnInit {
+    this.solutionText
+  ).subscribe({
+    next: () => {
+      this.solutionState = 'IN_PROGRESS';
```
magic String

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -213,4 +250,28 @@ export class ChallengeHeaderComponent implements OnInit {
+  onContinueChallenge(): void {
+  this.challengeStarted = true;
+  this.isEditorChallengeVisible = true;
+  this.status = 'IN_PROGRESS';
```
magic string

ivilarop commented on src/app/modules/challenge/components/challenge/challenge.component.ts:
```diff
@@ -38,15 +40,27 @@ export class ChallengeComponent implements OnInit, OnDestroy {
+  isFavorite: boolean = false
+  userSolution: UserSolution | null = null;
+  solutionText: string = '';
+  status: string = 'IN_PROGRESS'; 
```
magic String

ivilarop commented on src/app/modules/challenge/components/challenge/challenge.component.ts:
```diff
@@ -38,15 +40,27 @@ export class ChallengeComponent implements OnInit, OnDestroy {
   private readonly _authService = inject(AuthService)
+  private readonly solutionService = inject(SolutionService)
+  private userId: string = ''
+  public solutionState: 'NOT_STARTED' | 'IN_PROGRESS' | 'ENDED' = 'NOT_STARTED'
```
Same case as above. Check it out.

Silvitui replied:
The same for this. thank you

ivilarop commented on src/app/modules/challenge/components/challenge/challenge.component.ts:
```diff
@@ -55,33 +69,35 @@ export class ChallengeComponent implements OnInit, OnDestroy {
-              console.error('Error getting bookmarks:', err)
-            }
-          })
+   this._authService.getUserId().subscribe(userId => {
```
This snippet is not tabbed and is difficult to read. Also:

You are using literal strings ('IN_PROGRESS', 'ENDED', 'NOT_STARTED') that are already defined in the SolutionStatus enum.
Could you reuse it to maintain a single source of truth and avoid typing errors?

There are also several chained subscribe() methods without error handling or separation of responsibilities.
If we leave it this way for simplicity, it would be nice to at least:

Add an error to each subscribe or use catchError.

Type correctly with enums (SolutionStatus).

Add context to logs if included (console.error(...)).

Silvitui replied:
I've fixed the indentation to make it easier to read.
I’ve also replaced the string literals with SolutionStatus enum values to avoid typos.
For the observables, I’ve added proper error handlers and improved the console.error messages with more context.
Let me know if you see anything else to improve!

ivilarop commented on src/app/modules/challenge/components/challenge/challenge.component.ts:
```diff
@@ -92,11 +108,15 @@ export class ChallengeComponent implements OnInit, OnDestroy {
+  this.challengeStarted = started;
+  this.isEditorChallengeVisible = started;
+
+  this.status = started ? 'IN_PROGRESS' : 'NOT_STARTED';
```
magic string

Silvitui replied:
i’ve updated the code to use the SolutionStatus enum instead of the string literals. Thanks!

ivilarop commented on src/app/modules/challenge/components/challenge/challenge.component.ts:
```diff
@@ -126,6 +146,28 @@ export class ChallengeComponent implements OnInit, OnDestroy {
     })
   }
+onChallengeStart(): void {
+  this.challengeStarted = true;
```
Are so many boolean flags really necessary?

Silvitui replied:
in this method I'm already using challengeStarted to control the state.
I’ll review if solutionState and other flags like isEditorChallengeVisible are really needed, or if we can simplify them. Thanks!

ivilarop commented on src/app/modules/challenge/components/editor-challenge/editor-challenge.component.ts:
```diff
@@ -81,7 +92,8 @@ export class editorChallengeComponent implements OnInit, OnChanges, OnDestroy {
             EditorView.updateListener.of((update: ViewUpdate) => {
               if (update.docChanged) {
                 const content = this.editor.state.doc.toString()
-                localStorage.setItem('editorContent', content)
```
Is this change documented in the PR description?

I don't see anything in it about changes to the localStorage

Silvitui replied:
I’ve already documented this change in the description. Let me know if you think it needs to be clearer or if you'd like me to add more detail.

Silvitui committed:
Refactor challenge components to unify solution state handling and improve error management; fix button visibility and event communication; use SolutionStatus enum consistently to avoid magic strings

RonniedeTal commented:
I found some bugs in the challenge saving process, but it's working. Check it out.

-should be a continue button 
<img width="480" height="134" alt="Captura de pantalla 2025-07-14 122721" src="https://github.com/user-attachments/assets/10853800-ea10-47b7-bc67-221dc66bbcba" />

-when you continue the challenge appears this problem at the console, is working, take a little look
<img width="858" height="182" alt="Captura de pantalla 2025-07-14 123431" src="https://github.com/user-attachments/assets/c465a80d-4867-405c-9d2d-4d52de4be960" />

-when i start a challenge i have to press two times start challenge button to actually started and see the save button
1. 
<img width="977" height="187" alt="Captura de pantalla 2025-07-14 124855" src="https://github.com/user-attachments/assets/5ddf5eac-7d94-4c49-8d31-00b133cd7021" />
2. 
<img width="938" height="171" alt="Captura de pantalla 2025-07-14 124903" src="https://github.com/user-attachments/assets/323182c5-9b6a-433b-ba29-38e990d40385" />
3. 
<img width="954" height="146" alt="Captura de pantalla 2025-07-14 124910" src="https://github.com/user-attachments/assets/2a8ca902-62d8-4de2-b223-3768250de043" />

oscarrep reviewed (approved):
solutionStatus enum used correctly in all necessary places now!

Silvitui committed:
fix(challenge-header): fixed navigation to editor and "Start Challenge" button visibility without relying on localStorage

Silvitui committed:
resolving conflicts

Silvitui committed:
updating and fixing test and changelog

Silvitui committed:
covering test to pass sonarqube

Silvitui committed:
removing magic strings

Silvitui committed:
indentation in user-solution-status.enum.ts

Silvitui committed:
fix: show "Continue challenge" button and hide editor until clicked

oscarrep commented on et --hard ae3c640f:
remember to remove this unnecessary file

Silvitui replied:
I’ve already removed it, thanks!

oscarrep commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -8,6 +8,9 @@ import { SolutionService } from 'src/app/services/solution.service'
 import { ChallengeTab } from 'src/app/shared/enums/challenge-tab.enum'
 import { UserRole } from 'src/app/shared/enums/user-role.enum'
+import { SolutionStatus } from 'src/app/models/user-solution-status.enum'
+type SolutionState = SolutionStatus | 'NOT_STARTED';
```
should 'NOT_STARTED' go in the SolutionStatus enum?

Silvitui replied:
The backend doesn’t return a status like "NOT_STARTED", so we added it manually on the frontend.
We use it to know when a user has not started a challenge yet .This helps us show the "Start Challenge" button only when needed.

oscarrep replied:
okay! got it

ivilarop replied:
Should we include NOT_STARTED status in backend?

In the meantime, can we use enum instead of magic string?

oscarrep commented on src/app/modules/challenge/components/challenge-info/challenge-info.component.ts:
```diff
@@ -68,6 +72,12 @@ implements OnInit {
+  @Output() solutionChanged = new EventEmitter<string>();
+  @Input() languageId: string = '';
+ @Input() initialSolutionText: string = '';
+@Output() solutionChange = new EventEmitter<string>();
```
these declarations are not properly aligned. keeping consistent indentation helps improve code readability.

Silvitui replied:
I've fixed! thank you

oscarrep commented on src/app/modules/challenge/components/challenge/challenge.component.ts:
```diff
@@ -10,6 +10,12 @@ import { type Example } from 'src/app/models/challenge-example.model'
+import { SolutionService } from 'src/app/services/solution.service'
+import { UserSolution } from 'src/app/models/user-solution.interface'
+import { SolutionStatus } from 'src/app/models/user-solution-status.enum';
+type SolutionState = SolutionStatus | 'NOT_STARTED';
```
would including NOT_STARTED in the enum prevent the need to declare it here too?

Silvitui replied:
we keep NOT_STARTED as a separate string in a union type
This way, the enum matches only real backend values

oscarrep commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.html:
```diff
@@ -50,8 +50,11 @@ <h3 class="ms-2">{{title | dynamicTranslate}}</h3>
+class="btn btn-sm btn-primary"(click)="onContinueChallenge()">Continuar repte</button>
   <button
-  *ngIf="!challengeStarted && userRole !== USER_ROLE.ADMIN" class="btn btn-sm btn-primary" (click)="onStartChallenge()" >
+  *ngIf="solutionState === NOT_STARTED && userRole !== USER_ROLE.ADMIN" class="btn btn-sm btn-primary" (click)="onStartChallenge()" >
```
could we use 'solutionState === SolutionStatus.NOT_STARTED'?

Silvitui replied:
we can’t use SolutionStatus.NOT_STARTED because is not part of the SolutionStatus enum.

ivilarop replied:
The conversations say one thing and the code another.

Silvitui committed:
fixing indentation

oscarrep reviewed (approved):
Everything is good now

ivilarop commented on et --hard ae3c640f:
This file is still here

ivilarop commented on src/app/models/user-solution-status.enum.ts:
```diff
@@ -1,3 +1,4 @@
-    ENDED = "ENDED"
-  }
\ No newline at end of file
+  IN_PROGRESS = "IN_PROGRESS",
```
and the status NOT_STARTED ?

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.html:
```diff
@@ -50,8 +50,11 @@ <h3 class="ms-2">{{title | dynamicTranslate}}</h3>
   <div class="col-md-4 button d-none d-md-flex">
     <ng-container *ngIf="!solutionSent; else solutionSentMessage">
+       <button
+      *ngIf="solutionState === SolutionStatus.IN_PROGRESS && !challengeStarted && userRole !== USER_ROLE.ADMIN"
```
About your comment: As for the second part, I'll review the logic to see if we really need both conditions.

What conclusion did you reach?

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.spec.ts:
```diff
@@ -237,5 +231,158 @@ describe('ChallengeHeaderComponent', () => {
+  it('should set solutionState to NOT_STARTED if no matching solution is found', () => {
+    solutionService.fetchUserSolution.mockReturnValue(of([]));
+    component.loadUserSolutionStatus();
+    expect(component.solutionState).toBe('NOT_STARTED');
```
enum

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -20,13 +23,17 @@ export class ChallengeHeaderComponent implements OnInit {
   ) {}
-
+  public SolutionStatus = SolutionStatus;
+  public NOT_STARTED = 'NOT_STARTED';
```
enum

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -40,81 +47,94 @@ export class ChallengeHeaderComponent implements OnInit {
-      this.userRole = role
-    })
+      if (!solution) {
+        this.solutionState = 'NOT_STARTED';
```
enum

ivilarop commented on src/app/modules/challenge/components/challenge-header/challenge-header.component.ts:
```diff
@@ -40,81 +47,94 @@ export class ChallengeHeaderComponent implements OnInit {
-      this.activeId = ChallengeTab.SOLUTIONS // Mostrar botones de guardar y enviar solución
+        default:
+          console.warn(`Unhandled solution status: ${solution.status}`);
+          this.solutionState = 'NOT_STARTED';
```
enum

Silvitui committed:
feat: add NOT_STARTED as UI-only state to handle unstarted challenges
